### ISSUE TITLE
find a complete logical highway by name

### ISSUE THREAD

**Issue by jotpe** [drolbr/Overpass-API#95]
Currently there is no possibility to ask for a highway that belongs together with overpass-api. You can maybe ask for "Aachener Straße" in area[name=Köln] but you will receive three highways all over the city (and it gets broken at the city border).

For selecting only the desired logical highway, we can pick at first a piece of highway  with than name near a known point 50.937468375711916,
6.887891440378863. Now we need a new overpass magic, which asks recursive in in every radius around all way-nodes for additional highway parts.

Roland made already a QL-proposal, thanks

``` c
way
  (around:200,50.937468375711916,6.887891440378863)
  [highway]
  [name="Aachener Straße"];

complete->._
(
  ._; way(around:100)[highway][name="Aachener Straße"];
);
```

**Comment by mmd-osm** [comment]
You can try these sample queries now by temporarily changing your server in overpass turbo to http://dev.overpass-api.de/api_mmd/

If you come across some issue, please report it back here.

**Comment by sanderd17** [comment]
Hmm, just got here from that diary post (http://www.openstreetmap.org/user/Chenshi/diary/35949). And I'm a bit confused with the syntax. AFAICS, complete just seems to be a foreach loop, where the input set can be added to on the fly, and that also has an output.

By requiring the line `._; > ->.a;`, it becomes apparent that you need to manually get the nodes from the elements, and use those nodes explicitly to see if other ways are connected. While that would be something I inherently expect in a `complete` statement.

Modifying the foreach definition so sets can be adapted while adding to it will add about the same functionality AFAICS, while it wouldn't introduce an extra keyword. Something like this could then also give all connected ways to some selection, while being hardly more cumbersome:
```C
area["name"="Roeselare"]["admin_level"=8]->.a;
way["highway"="primary]"["name"="Diksmuidsesteenweg"](area.a)->.ways;
foreach.ways 
(  
  > ->.nodes; 
  (
    .ways;
    way(bn.nodes)["highway"="primary"](area.a);
  )->.ways; 
);
.ways out geom;
```
This is a common practice in at least some programming languages, so it's purely a design choice. Compare it to JavaScript, where the following will be an infinite loop (it will at least loop until it runs out of integers):

```javascript
var set = new Set([1]);
for (var v of set)
    set.add(v+1);
```

On the other hand, if `foreach` is left as it currently is, and `complete` is added, getting the nodes should be part of the iteration IMO, so it would become something like

```C
[bbox:{{bbox}}]
 way(28742708)->.way;
 complete.way->.nodes( way(bn.nodes)[railway]; )->.result;
.result out geom;
```
or without sets
```C
[bbox:{{bbox}}]
way(28742708);
complete( way(bn)[railway]; );
out geom;
```

**Comment by mmd-osm** [comment]
> complete just seems to be a foreach loop, where the input set can be added to on the fly, and that also has an output.

That's not quite accurate. While `foreach` processes the elements one by one, `complete` provides the complete inputset to the statements inside. They're free to do what ever they want, but at the end of the day, they have to produce some outputset. `complete` then checks, if the iteration produced any new elements. If there are some new elements, a new iteration will be triggered just with those new elements. On the other hand, if the inner statements didn't produce any new elements, `complete` is done and returns the set of all elements of all iterations back as a result.

**Comment by mmd-osm** [comment]
(Note: I updated my previous answer quite a bit, as I didn't have much time earlier today. Not sure, what kind of GitHub are triggered...)

**Comment by mmd-osm** [comment]
Combined with 0.7.54, we can now find a certain number of adjacent ways:

```
(
  way(143101141);
  complete(
      (.k; make x -> .f ;) -> .k;
      way._(if:k.count(deriveds) <= 15);
      way[highway][highway!~"^(path|track|cycleway|footway)$"](around:0);
  );
);  

out geom;
```

http://overpass-turbo.eu/s/nRp

**Comment by mmd-osm** [comment]
Another example: Auto-complete way with close-by highways of the same name:

```C
( way[name](473306924);
  complete (
    way[name](around:500)(if:t[name] == u(t[name]));
  );
);                   
out geom;
```

http://overpass-turbo.eu/s/o2C

**Comment by drolbr** [comment]
Available by pull request integration in 409c6a5349d50b535b43d170e8fd5f9ebfac287f

**Comment by vin-ni** [comment]
The Demo is down unfortunately.
How would I get all streets of Berlin as one object per street?
